### PR TITLE
Add ^NumExpr preprocessing directive to perform arithmetic operations on YAML configuration files

### PR DESCRIPTION
**Pull request by ykangpark**
Many times it would be nice to have arithmetic with parameters in YAML files.

This is a proposal to add this functionality. Not sure if the syntax is the most appropriate, since it might be confusing with doc generators. Given that ^Parameter also follows this ^xxxx style, I used the same.

### CODE REVIEW

**Review by aggrand — COMMENTED** [review]
I don't have too strong of an opinion to this, if people would find it useful! That said, I feel like the desire for this points to a deeper need for more flexible workload configuration.

also cc @rtimmons  who might have opinions on new workload-level syntax!

File: src/lamplib/src/genny/tasks/preprocess.py [hunk 1]
```diff
@@ -279,6 +282,44 @@ def _replace_param(self, input):
...
+        if "Expr" not in input:
+            msg = "Invalid keys for '^NumExpr', please set 'Expr'" f" in following node: {input}"
+            raise ParseException(msg)
+
+        if type(input["Expr"]) != str:
+            msg = "Invalid value for 'Expr', which must be a string," f" in following node: {input}"
+            raise ParseException(msg)
+
+        parsedDict = {}  # Pass empty dict to avoid yaml to access context of this function
+        if "Dict" in input:
```
Comment (aggrand) [inline src/lamplib/src/genny/tasks/preprocess.py:295]: Could we call this "args" or "values"? I'm not sure it's obvious how a dict would necessarily be used.

**Review by rtimmons — CHANGES_REQUESTED** [review]
I'm not quite sure why this is necessary and fear as Raiden suggested that this might be papering over other syntax shortcomings. This said, I think it makes sense to see where this goes based on actual usage and to consider further syntax changes with real-world usage in mind.

A couple things here:

1. Please use `snake_case`
2. The `Dict` name is ambiguous. Consider `{^Expr: {withExpression:"a - b", andValues:{a:1,b:2}}`. We haven't been super rigorous with our syntax, so it will be mostly subjective. My point is to make the name at least somewhat meaningful.
3. A test-case that shows this is evaluated recursively and/or potentially within the context of a value-generator would be useful for regressions e.g. `{Document: {^FastRandomString:{length:{^Expr:...}}`.
4. We have a special `^Verbatim` syntax that allows for values to be passed through without any `^`-handling. I don't know if the existing Python is aware of this, but if so please ensure an expression like `{^Verbatim: {^Expr:...}}` gets evaluated to literal `{^Expr:...}` in the preprocessor output.

Otherwise this code looks straightforward and well-tested so not a lot of change to get this over the finish line. Please push back or ping back with any questions or disagreements.

**Review by aggrand — CHANGES_REQUESTED** [review]
Mostly had a nitpick, and also a request that an example of this be added to a HelloWorld.yml workload in `src/workloads/docs`.

Also a point of discussion, this is getting pretty close to just evaluating arbitrary python. Is there an argument for going all-in on that to pre-empt the next time something like this is needed?

File: src/lamplib/src/genny/tasks/preprocess.py (hunk 1 above)
Comment (ykangpark) [reply src/lamplib/src/genny/tasks/preprocess.py]: Thanks for the feedback. Please see latest commits where this has been changed. Also added an example to HelloWorld.yml

**Comment by ykangpark** [comment]
> I'm not quite sure why this is necessary and fear as Raiden suggested that this might be papering over other syntax shortcomings. This said, I think it makes sense to see where this goes based on actual usage and to consider further syntax changes with real-world usage in mind.
> 
> A couple things here:
> 
> 1. Please use `snake_case`
> 2. The `Dict` name is ambiguous. Consider `{^Expr: {withExpression:"a - b", andValues:{a:1,b:2}}`. We haven't been super rigorous with our syntax, so it will be mostly subjective. My point is to make the name at least somewhat meaningful.
> 3. A test-case that shows this is evaluated recursively and/or potentially within the context of a value-generator would be useful for regressions e.g. `{Document: {^FastRandomString:{length:{^Expr:...}}`.
> 4. We have a special `^Verbatim` syntax that allows for values to be passed through without any `^`-handling. I don't know if the existing Python is aware of this, but if so please ensure an expression like `{^Verbatim: {^Expr:...}}` gets evaluated to literal `{^Expr:...}` in the preprocessor output.
> 
> Otherwise this code looks straightforward and well-tested so not a lot of change to get this over the finish line. Please push back or ping back with any questions or disagreements.

Thanks for the feedback. 
I addressed all points except point 4, given that ^Verbatim is processed at DocumentGenerator.cpp and the Python is not aware.

**Review by rtimmons — APPROVED** [review]

**Pull request merged**